Fix based on comments

xbauch opened the pull request.

NlightNFotis commented on src/goto-harness/function_call_harness_generator.h:
```diff
@@ -25,17 +25,17 @@ class function_call_harness_generatort : public goto_harness_generatort
+  void validate_options(const goto_modelt &goto_model) override;
 
 private:
+  irep_idt harness_fun_name;
```
I'm not sure if this should go here? I think similar stuff is in `implt`.

xbauch replied:
True, this shouldn't be here.

NlightNFotis commented on src/goto-harness/function_call_harness_generator.h:
```diff
@@ -25,17 +25,17 @@ class function_call_harness_generatort : public goto_harness_generatort
     const std::list<std::string> &values) override;
 
-  void validate_options() override;
+  void validate_options(const goto_modelt &goto_model) override;
```
Why are all of these suddenly passed a `goto_model`? And especially `validate_options` shouldn't need one?

xbauch replied:
Michael T wanted checks to happen sooner. For example checking that the function name exists in the symbol table. So I moved the checks to validate_options.

xbauch replied:
I guess the question is: Is `validate_options` the right place for the checks?

hannes-steffenhagen-diffblue replied:
@xbauch It's a reasonable thing to do, we've had this discussion with @danpoe already.

NlightNFotis commented on src/goto-programs/goto_program.h:
```diff
@@ -539,8 +539,11 @@ class goto_programt
+    auto location_target = std::next(
       instructions.begin(), location_number - start_location_number);
+
+    CHECK_RETURN(location_target->location_number == location_number);
```
I think this was one of the suggestions by Michael. I think he meant this should be done before `std::next` happens, as it can crash without the warrantied property this is assuming (that locations are contiguous).

xbauch replied:
https://github.com/diffblue/cbmc/pull/4150/files#r255405529 .. I think `std::next` would only crash if we wanted it to move the iterator beyond `end()`.

NlightNFotis commented on src/goto-harness/memory_snapshot_harness_generator.cpp:
```diff
@@ -232,57 +276,19 @@ void memory_snapshot_harness_generatort::get_memory_snapshot(
-  code_blockt harness_function_body;
-
-  add_assignments_to_globals(snapshot, harness_function_body);
+  code_blockt harness_function_body = add_assignments_to_globals(snapshot);
```
I think this could be better with a comment to explain intent. The name and signature are not immediately obvious what they do.

NlightNFotis commented on src/goto-harness/memory_snapshot_harness_generator.cpp:
```diff
@@ -62,114 +64,154 @@ void memory_snapshot_harness_generatort::handle_option(
   {
     const symbolt &symbol = pair.second;
 
     if(!symbol.is_static_lifetime)
```
If I understand it correctly, this is checking for a global right? I think this required one more property but I'm not sure which one it was (I think `is_lvalue` or something). I'm not sure if this matters here.

sonodtt replied:
Hmm here we seem to want to avoid initialising non-globals.
seconded "is_lvalue" in the case of globals though - I recollect this from Fotis & Hannes' recent Harness PR. 

However as these variables come from the snapshot - it may well be that the symbol table is relatively 'clean' - but is that a safe assumption?

In the generic case this seems to be how to check for globals:
"             symbol.is_static_lifetime && symbol.is_lvalue &&
             symbol.type.id() != ID_code &&
             !has_prefix(id2string(symbol.name), CPROVER_PREFIX))"

For non-globals it may we be that !symbol.is_static_lifetime is sufficient. 
Hmm. Worth a think.

NlightNFotis reviewed (commented):
Good work. Some comments. Sorry I can't verify right now my questions. No blockers though, just questions.

sonodtt commented on src/goto-programs/goto_program.h:
```diff
@@ -539,8 +539,11 @@ class goto_programt
+    auto location_target = std::next(
       instructions.begin(), location_number - start_location_number);
+
+    CHECK_RETURN(location_target->location_number == location_number);
```
You added this based on this comment: "I'm not sure we provide a guarantee that location numbers are contiguous. You might want to add a CHECK_RETURN to make sure what you are about to return has the expected location_number."

possible comment:
"location numbers may not be contiguous. Verify expected location number holds:

If who wrote this did not know it -  then who reviews it or uses it in the future may also not. So clarify why you added this or they will be looking at it going - "hmm... why?"

Take these as opportunities to provide clarity - not just fixes. 

*Pass on what you know or have found out... *

xbauch replied:
Added.

sonodtt commented:
Quite a few changes / fixes. Some comments - but not many.

For the fixes - many of which will be non-obvious (or they would not have been required) - please provide comments. 

*pass on what you know or have found out...*

sonodtt reviewed (changes requested):
Quite a few changes / fixes.  In some cases you have provided great information (e.g. cprover_initialize) - many thanks!

For the fixes - some are non-obvious (or they would not have been required) - please provide comments in such circumstances.

hannes-steffenhagen-diffblue commented on src/goto-harness/goto_harness_generator.h:
```diff
@@ -47,6 +48,8 @@ class goto_harness_generatort
     const std::string &option,
     const std::list<std::string> &values);
+
+  irep_idt harness_function_name;
```
This change seems unnecessary, I'd prefer you didn't do that.

hannes-steffenhagen-diffblue commented on src/goto-harness/memory_snapshot_harness_generator.cpp:
```diff
@@ -62,114 +64,154 @@ void memory_snapshot_harness_generatort::handle_option(
-  auto ins_it3 = goto_program.insert_after(ins_it2);
-  ins_it3->make_goto(goto_program.const_cast_target(it));
+  // initialise func_init_done_var in __CPROVER_initialize if it is present
+  // so that it's FALSE value is visible before the harnessed function is called
```
That still doesn't make a lot of sense to me - you should be able to initialise it in the harness function without modifying `__CPROVER_initialize` (or indeed assuming there even _is_ a `__CPROVER_initialize`)?

danpoe reviewed (approved):
LGTM

Commit:
Search the JSON for symbol table

In case we get an array (and not an object) we first find the right object and
then pass it to be converted.

xbauch commented:
@danpoe I swapped the commits and moved the `harness_function_name` check to `doit` as per @hannes-steffenhagen-diffblue suggestion. Should be ready now.

NlightNFotis reviewed (approved):
LGTM

Commit:
Fix based on comments
